File: Python/Systems of linear equations/LU Decomposition/Doolittle Decomposition/lu_doolittle.py
import numpy as np

def lu_decomposition(A):
    # Perform Doolittle LU decomposition of matrix A
    n = len(A)
    L = np.zeros((n, n))
    U = np.zeros((n, n))

    for i in range(n):
        for j in range(i, n):
            # Compute upper triangular matrix U
            sum = 0
            for k in range(i):
                sum += L[i][k] * U[k][j]
            U[i][j] = A[i][j] - sum

        for j in range(i, n):
            # Compute lower triangular matrix L
            if i == j:
                L[i][j] = 1
            else:
                sum = 0
                for k in range(i):
                    sum += L[j][k] * U[k][i]
                L[j][i] = (A[j][i] - sum) / U[i][i]

    return L, U

# Example usage
A = np.array([[4, 6, 10], [6, 25, 19], [10, 19, 62]])

L, U = lu_decomposition(A)

# Print L and U matrices
print("L matrix:")
print(L)
print("U matrix:")
print(U)

File: Matlab/Systems of linear equations/LU Decomposition/Doolittle Decomposition/lu_doolittle.m
function [L, U] = lu_doolittle(A)
    % Function to perform Doolittle's LU decomposition
    % Input:
    %   A - square matrix (n x n)
    % Output:
    %   L - Lower triangular matrix with unit diagonal
    %   U - Upper triangular matrix

    % Get the size of the matrix
    [n, m] = size(A);

    % Ensure the matrix is square
    if n ~= m
        error('Matrix A must be square.');
    end

    % Initialize L and U matrices
    L = eye(n); % Lower triangular matrix with 1's on the diagonal
    U = zeros(n); % Upper triangular matrix

    % Perform decomposition
    for j = 1:n
        % Compute U (row by row)
        for i = 1:j
            sumU = 0;
            for k = 1:i-1
                sumU = sumU + L(i,k) * U(k,j);
            end
            U(i,j) = A(i,j) - sumU;
        end

        % Compute L (column by column)
        for i = j+1:n
            sumL = 0;
            for k = 1:j-1
                sumL = sumL + L(i,k) * U(k,j);
            end
            L(i,j) = (A(i,j) - sumL) / U(j,j);
        end
    end
end

% To test the written function, run the following code in the MATLAB
% command window (be sure that the M-File is in the current folder)
% A = [4 6 10; 6 25 19; 10 19 62];
% [L, U] = lu_doolittle(A);
% disp(L)
% disp(U)